index.js

```diff
@@ -67,25 +67,30 @@ const rrdir = module.exports = async function* (dir, opts = {}, {includeMatcher,
     const path = makePath(dirent, dir, encoding);
     if (excludeMatcher && excludeMatcher(encoding === "buffer" ? String(path) : path)) continue;
 
+    const isSymbolicLink = opts.followSymlinks && dirent.isSymbolicLink();
+
     let stats;
-    if (opts.stats) {
-      try {
-        stats = await (opts.followSymlinks ? stat : lstat)(path);
-      } catch (err) {
-        if (opts.strict) throw err;
-        yield {path, err};
+    if (!includeMatcher || includeMatcher(encoding === "buffer" ? String(path) : path)) {
+      if (opts.stats || isSymbolicLink) {
+        try {
+          stats = await (opts.followSymlinks ? stat : lstat)(path);
+        } catch (err) {
+          if (opts.strict) throw err;
+          yield {path, err};
+        }
       }
+
+      yield build(dirent, path, stats, opts);
     }
 
     let recurse = false;
-    if (opts.followSymlinks && dirent.isSymbolicLink()) {
+    if (isSymbolicLink) {
       if (!stats) try { stats = await stat(path); } catch {}
```

Comment:
I think these lines can be removed because `stats` is guaranteed to exist when `isSymbolicLink` is true.

Reply:
Actually, no I think these need to stay because `lstat` vs `stat` differences.